### ISSUE TITLE
[Inductor] functorch_dp_cifar10 training accuracy has failed 100% on H100 since 2026-06-13 (all configs)

### ISSUE THREAD

**Issue by atalman** [pytorch/pytorch#195483]
### 🐛 Describe the bug

`functorch_dp_cifar10` **training** accuracy has been failing 100% of the time on H100 in the inductor perf-nightly dashboard since **2026-06-13**, across every config (`default`/`no_cudagraphs`, `with_cudagraphs`, `dynamic`, `cpp_wrapper`, `deterministic_perf`). It has not passed once on H100 in ~11 weeks.

The verdict is `fail_accuracy` (numeric divergence from eager), not `fail_to_run`.

This was found while comparing 2.13.0 vs 2.14.0 release perf numbers. It is **not** a release-branch regression — it landed on `main` in June and the 2.14 branch inherited it. The 2.13.0 branch was cut before the offending range, which is why 2.13.0 looks clean by comparison.

### Evidence

H100, `refs/heads/main`, training accuracy, from `benchmark.oss_ci_benchmark_torchinductor`:

| Weeks | Pass | Fail | Fail rate |
|---|---|---|---|
| 2026-03-01 → 2026-06-12 (14 wks) | 524 | 0 | **0%** |
| 2026-06-13 → 2026-08-31 (12 wks) | 0 | 410 | **100%** |

Daily, the flip is a single step with all configs moving together:

| Day | head_sha | Result |
|---|---|---|
| 2026-06-11 | `ce9267a458` | 5 / 5 pass |
| 2026-06-12 | `083e2617b8` | 5 / 5 pass |
| **2026-06-13** | **`2637e3bdb1`** | **0 / 5 pass** |
| 2026-06-14 | `27f34a3e20` | 0 / 6 pass |
| 2026-06-15 | `6a231d0d3e` | 0 / 5 pass |

### Bisect range

`083e2617b8` (last good) → `2637e3bdb1` (first bad) — [59 commits](https://github.com/pytorch/pytorch/compare/083e2617b8...2637e3bdb1).

Plausible suspects in range, given this model is Opacus DP-SGD (per-sample gradients via `functorch` + injected Gaussian noise), unverified:

- `022ae6616f` — [Preserve metadata for fused Inductor RNG nodes (#184316)](https://github.com/pytorch/pytorch/pull/184316) — model depends on RNG for the DP noise term
- `8bc0aa73bd` — [Fix AOTAutograd wrapper subclass metadata mutation (#186515)](https://github.com/pytorch/pytorch/pull/186515) — per-sample gradients go through tensor subclasses
- `ca94d752ae` — [cuDNN Upgrade 12.8, 13.0 (and 12.9) wheels to cuDNN 9.23.1.3 (#187091)](https://github.com/pytorch/pytorch/pull/187091) — would explain why this is CUDA-arch-sensitive
- `5b85fbbc6b` — [Split very large reductions out of combo kernel partitioning (#186668)](https://github.com/pytorch/pytorch/pull/186668)
- `84eb6a3d64` — [Fix Dynamo random.shuffle and random.sample (#186499)](https://github.com/pytorch/pytorch/pull/186499)

### Scope

- **Training only.** Inference is unaffected: 4960 / 4960 pass over the same 90-day window.
- **Arch-sensitive.** Weekly fail rate on `main` since June:

| GPU | Pattern |
|---|---|
| H100 | 0% → **100% from 2026-06-13, sustained** |
| A100 | 0% until Jun 21, ~76% Jun 28 – Aug 9, back to **0% since Aug 16** |
| B200 | Oscillates 0% ↔ 100% (already 100% in May, predates this) |
| MI300X / MI325X | Oscillates |
| AMD Radeon | ~3% baseline |
| Intel XPU | 0 / 258 |

The non-H100 numbers suggest this model also sits close to its accuracy tolerance in general, which is likely why a 100%-failing config went unnoticed for 11 weeks — it blends into pre-existing noise, and perf-nightly accuracy is not a blocking signal.

### Repro

Derived from `.ci/pytorch/test.sh` (`test_perf_for_dashboard`, `default-true` / accuracy target):

```bash
python benchmarks/dynamo/torchbench.py \
  --accuracy --no-translation-validation \
  --training --amp --backend inductor \
  --disable-cudagraphs --device cuda \
  --only functorch_dp_cifar10
```

### Versions

`main` @ `083e2617b8` (good) and `2637e3bdb1` (bad); reproduced continuously on `main` through 2026-08-31. Device: NVIDIA H100 80GB HBM3, dtype `amp`.

Data source: `benchmark.oss_ci_benchmark_torchinductor`, filtered to `head_branch='refs/heads/main'`, `metric_name='accuracy'`, `benchmark_mode='training'`, perf-nightly harness only (`benchmark_extra_info['output'] LIKE '%test-reports/inductor\_%'`), deduplicated with `argMax(value, timestamp)`.

<sub>Analysis performed with Claude.</sub>

cc @ezyang @gchanan @kadeng @msaroufim @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @muchulee8 @amjames @aakhundov @coconutruben @jataylo

**Comment by atalman** [comment]
Correction / refinement to the suspect list above, prompted by #186613.

**#186613 ("Use larger small-tensor multiplier for functorch_dp_cifar10") never landed.** It was closed 2026-06-18 after a failed `@pytorchbot rebase` on 06-16. `functorch_dp_cifar10` is **not** in `require_larger_multiplier_for_smaller_tensor` on `main` today — that list still contains only `yolov3`.

That PR was compensating for #186379 (`[inductor] Don't fold float x * 0 to 0 in uniform-value constant folding`), which has a land / revert / reland history that matters here. Tracking the guard line `tensor_val.dtype.is_floating_point or tensor_val.dtype.is_complex` in `torch/_inductor/fx_passes/joint_graph.py`:

| Commit | Date | Guard present |
|---|---|---|
| `eb80dd583a` (#186379) | 2026-06-06 | yes — landed |
| `9cf6d4fe76` | 2026-06-10 | **no — reverted** |
| `083e2617b8` (last good) | 2026-06-12 | no |
| **`2637e3bdb1` (first bad)** | **2026-06-13** | **no** |
| `ab9fa1b295` (#187580) | 2026-06-18 | yes — relanded |
| `main` | today | yes |

Ancestry is linear across these (`9cf6d4fe76...2637e3bdb1` is ahead 174 / behind 0; `2637e3bdb1...ab9fa1b295` ahead 262 / behind 0).

**So the constant-folding change was out of the tree on the day H100 flipped**, and H100 was green for the whole Jun 6–10 window when it *was* in. It therefore does not explain the 06-13 regression, and it should be excluded from the 59-commit range hypotheses. Whatever williamwen42 was chasing in early June was a different manifestation, presumably on another config or device.

This leaves two likely-distinct problems:

1. **The 06-13 H100 regression** — still unexplained, still inside [`083e2617b8...2637e3bdb1`](https://github.com/pytorch/pytorch/compare/083e2617b8...2637e3bdb1). The suspects in the issue body stand, minus any constant-folding theory.
2. **From 06-18 onward, `main` carries the relanded #187580 without the tolerance entry that #186613 judged necessary.** Because H100 has been saturated at 100% failure since 06-13, this second effect is unobservable in the dashboard — it cannot be separated from (1) by looking at pass/fail.

Practical consequence: **reviving #186613 is likely necessary but not sufficient.** It addresses the borderline-RMSE issue on the 128-element BatchNorm bias gradient that the relanded PR reintroduces, but it cannot account for a failure that began five days before the reland. Both need handling, and fixing only the tolerance may make this look green while leaving the 06-13 cause in place.

cc @williamwen42 (author of #186613)

<sub>Analysis performed with Claude.</sub>

**Comment by williamwen42** [comment]
Bisected this on H100. **There are two independent causes, and neither is a numerics degradation.** The 59-commit range in the issue body can be fully exonerated.

## TL;DR

| | cause | lands | fails at | ratio |
|---|---|---|---|---|
| **#1** — the Jun 13 flip | `ca94d752ae` (#187091), CI-only cuDNN bump 9.20.0.48 → 9.23.1.3 | 2026-06-12 | `layer4.1.bn2.weight` | 7.57 |
| **#2** — separate, later | `8003932681` (#191945), `[CUDA] Add NHWC GroupNorm Implementation` | 2026-08-14 | `layer4.0.bn2.bias` | 3.54 |

On current `main` both are live and **each is independently sufficient** — fixing only one leaves this red.

## Effect #1: the Jun 13 flip is the cuDNN bump, not any source commit

`ca94d752ae` was already suspect #3 in the issue body ("would explain why this is CUDA-arch-sensitive"). It's confirmed. It touches only `.ci/` and `.github/` — **zero source files** — so a from-source bisect can never separate on it. Both endpoints build and pass identically on a local machine, which is exactly what we saw first.

Holding the source fixed and moving only the cuDNN runtime (`LD_LIBRARY_PATH`, verified via `torch._C._cudnn.getRuntimeVersion()`):

| torch commit | cuDNN 9.9.0 | cuDNN 9.20.0.48 | cuDNN 9.23.1.3 |
|---|---|---|---|
| `083e2617b8` (last good) | pass (1.84) | pass (1.84) | **fail_accuracy (7.57)** |
| `2637e3bdb1` (first bad) | pass (1.84) | pass (1.84) | **fail_accuracy (7.57)** |
| `ac5cb186e0` (`main`, Sep 1) | fail (3.54) | fail (3.54) | **fail (7.56)** |

(ratio = max `res/ref` RMSE across all compared tensors; multiplier is 3.0 for the failing ones)

**The last-good commit `083e2617b8` also fails under 9.23.1.3.** No commit in `083e2617b8..2637e3bdb1` is necessary to produce the failure, so the source-level suspects — #184316, #186515, #186668, #186499 — are all ruled out.

The config chain for why CI's cuDNN changed on that date:
- the H100 job builds in `ci-image:pytorch-linux-jammy-cuda13.0-cudnn9-py3-gcc11-inductor-benchmarks` (`inductor-perf-test-nightly-h100.yml`)
- that image's cuDNN comes from `.ci/docker/common/install_cuda.sh::install_130`
- `ca94d752ae` bumps `install_130` (and `install_128/129/132`) `CUDNN_VERSION=9.20.0.48` → `9.23.1.3`
- the image tag is `git rev-parse HEAD:.ci/docker` (`docker-builds.yml`), and that tree hash rotates at exactly this commit: `8e94cc61` → `ec74f8e7`

## Effect #2: `8003932681` (#191945), NHWC GroupNorm

Bisecting `2637e3bdb1..ac5cb186e0` (3780 commits) with cuDNN pinned to **9.20.0.48** to hold effect #1 off:

| commit | date | verdict | ratio |
|---|---|---|---|
| `e2814d0069` (direct parent) | Aug 14 | pass | 1.84 |
| **`8003932681`** (#191945) | Aug 14 | **fail_accuracy** | **3.54** |

This is causally tied to the model rather than coincidental: `functorch_dp_cifar10` patches ResNet-18's norm layers to GroupNorm while keeping the `bn*` attribute names —

```python
# Generate a resnet18, patch the BatchNorm layers to be GroupNorm
norm_layer=(lambda c: nn.GroupNorm(min(c, 32), c))
```

— so the `bn2` that fails **is** a GroupNorm, and #191945 rewrites the CUDA GroupNorm kernel plus its decomposition and ref.

This landed while the dashboard was already saturated at 100% failure, so it produced no visible change. That's the unobservable second effect predicted in the comment above — though it turns out to be GroupNorm, not the #187580 constant-folding reland. #187580 is separately ruled out: its fold never fires on this model (verified by instrumenting the fold body, zero hits).

## Both are tolerance artifacts, not accuracy regressions

At the #191945 boundary, **both** paths got *more* accurate — eager just improved much more, so the ratio crossed the fixed multiplier:

| | inductor vs fp64 | eager vs fp64 | ratio |
|---|---|---|---|
| parent | 3.832e-3 | 2.273e-3 | 1.69 |
| #191945 | 1.189e-3 | 3.356e-4 | **3.54** |

The cuDNN case has the same shape: eager's error drops to 3.9e-5 while inductor's stays ~2.9e-4 → 7.57. In both cases an eager-side kernel got better and inductor's path didn't follow. That argues against the `module: correctness (silent)` framing.

## A trap for the obvious fix

Both failing tensors are **512 elements**. `get_multiplier()` gives 10.0 for `numel <= 10`, 8.0 for `numel <= 500`, then falls through to 3.0 — so at 512 it yields **3.0 regardless**. Adding `functorch_dp_cifar10` to `_require_larger_multiplier_for_smaller_tensor` (the #186613 approach) **would not fix either failure**. #186613 targeted a 128-element BN bias, which is a third, distinct tensor.

## Caveats

- Reproduced on a local CUDA 12.8/12.9 build against system cuDNN 9.9.0 headers, swapping only the runtime `.so`. **CI builds CUDA 13.0.** The cuDNN A/B is clean in this config but has not been reproduced on a CUDA 13.0 build.
- We have not read a Jun 12 vs Jun 13 CI job log confirming the installed cuDNN version in situ; the chain above is read off the repo plus the local A/B.
- `main` and all bisect steps were built with GCC 14 and `USE_FBGEMM=0` (an Aug-11 fbgemm snapshot doesn't compile otherwise; fbgemm is unreachable from this AMP/CUDA path). This is controlled at the bisect boundary — parent and child share the config — but the June→Sep cross-comparison is looser. Re-running June under GCC 14 reproduced 1.84 / 7.57 exactly.
- The A100 pattern in the issue body (0% until Jun 21, ~76% Jun 28–Aug 9, 0% since Aug 16) does not fit a pure-cuDNN story cleanly. Arch-specific kernel selection plus this model sitting near tolerance would explain it, but that is a hypothesis, not measured.

cc @eqy (author of both #187091 and #191945), @atalman, @ngimel

<sub>Analysis performed with Claude.</sub>

**Comment by eqy** [comment]
Is there a human note? 
```
Both are tolerance artifacts, not accuracy regressions
At the #191945 boundary, both paths got more accurate — eager just improved much more, so the ratio crossed the fixed multiplier:

inductor vs fp64	eager vs fp64	ratio
parent	3.832e-3	2.273e-3	1.69
#191945	1.189e-3	3.356e-4	3.54
The cuDNN case has the same shape: eager's error drops to 3.9e-5 while inductor's stays ~2.9e-4 → 7.57. In both cases an eager-side kernel got better and inductor's path didn't follow. That argues against the module: correctness (silent) framing.
```
sounds like it is not a correctness issue?

**Comment by williamwen42** [comment]
I reviewed the agent output before posting. It would be good to independently confirm that the 2 blamed PRs could have reasonably caused this apparent regression, and if so, then we should just raise the tolerance.

**Comment by atalman** [comment]
Removing this after discussion with @zou3519 not criciaal for 2.14.1, tolerance issue
